src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = Object.assign({}, state);
+  const states = [];
+
+  actions.forEach((action) => {
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    } else if (action.type === 'clear') {
+      currentState = {};
+    }
```

Comment:
Instead of using multiple `if` statements, a `switch` statement would be more appropriate here as you have a limited amount of conditions to check for. This will make your code cleaner and more maintainable. Remember to include a `default` case for error handling, as per checklist items 2 and 3.